Assignment/PingProcess.cs

```diff
@@ -0,0 +1,229 @@
+﻿using System.ComponentModel;
+using System.Diagnostics;
+using System.Text;
+
+namespace Assignment;
+
+public record struct PingResult(int ExitCode, string? StdOutput, string? StdError);
+
+public class PingProcess
+{
+    private ProcessStartInfo StartInfo { get; } = new("ping");
+
+    public PingResult Run(string hostNameOrAddress, IProgress<string?>? progress = null)
+    {
+
+        string pingArg = Environment.OSVersion.Platform is PlatformID.Unix ? "-c" : "-n";
+
+        StartInfo.Arguments = $"{pingArg} 4 {hostNameOrAddress}";
+        StringBuilder? stdOutput = null;
+        StringBuilder? stdError = null;
+        void updateStdOutput(string? line)
+        {
+            (stdOutput ??= new StringBuilder()).AppendLine(line);
+            if (progress is not null) progress?.Report(line);
+        }
+        void updateStdError(string? line) =>
+            (stdError ??= new StringBuilder()).AppendLine(line);
+        Process process = RunProcessInternal(StartInfo, updateStdOutput, updateStdError, default);
+        return new PingResult(process.ExitCode, stdOutput?.ToString(), stdError?.ToString());
+
+    }
+
+    // Task 1]
+    public Task<PingResult> RunTaskAsync(string hostNameOrAddress)
+    {
+        return Task.Run(() => Run(hostNameOrAddress));
+
+    }
+
+    // Tasks 2 & 3]
+    async public Task<PingResult> RunAsync(
+    string hostNameOrAddress, CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            PingResult result = await Task.Run(() => Run(hostNameOrAddress), cancellationToken);
+            return result;
+        }
+        catch (OperationCanceledException)
+        {
+            throw new AggregateException(new TaskCanceledException());
+        }
+
+    }
+
+    // Task 4]
+    async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)
+    {
+        int count = 0;
+        StringBuilder stringBuilder = new StringBuilder();
+        var semaphore = new SemaphoreSlim(1);

```

Comment:
```suggestion
        var semaphore = new SemaphoreSlim(1,1);

```

While semaphore's aren't necessary here, you should always declare a maximum if you only want 1 thread to enter in at a time.